src/client/client_impl.cc

```diff
@@ -79,6 +82,189 @@ google::protobuf::util::Status ClientImpl::CreateFile(
+
+    // Issue OpenFileReply rpc and check stastus
+    StatusOr<OpenFileReply> open_file_or(
+          master_metadata_service_client_->SendRequest(
+              open_file_request, client_context));
+
+    // Handle error
+    if (!open_file_or.ok()) {
+      LOG(ERROR) << "OpenFileRequest failed due to " 
+                 << open_file_or.status().error_message();
+      return open_file_or.status();
+    }
+
+    // Cache file chunk metadata
+    auto open_file_reply(open_file_or.ValueOrDie());
+    cache_file_chunk_metadata(filename, chunk_index, open_file_reply);
+  
+    chunk_handle = cache_manager_->GetChunkHandle(filename, chunk_index)
+                                       .ValueOrDie();
+    chunk_version = cache_manager_->GetChunkVersion(chunk_handle).ValueOrDie();
+    chunk_server_location_entry = 
+        cache_manager_->GetChunkServerLocation(chunk_handle).ValueOrDie();
+  }
+
+  // If no locations are there, we need to return error
+  if (chunk_server_location_entry.locations.empty()) {
+    return google::protobuf::util::Status(
+               google::protobuf::util::error::UNAVAILABLE,
+               "No chunk server has been found");
+  }
+
+  // Loop over all server locations but return immediately as long as one 
+  // read request is successful
+  for (auto& location : chunk_server_location_entry.locations) {
+    // Prepare for the read file chunk request
+    ReadFileChunkRequest read_file_chunk_request;
+    read_file_chunk_request.set_chunk_handle(chunk_handle);
+    read_file_chunk_request.set_chunk_version(chunk_version);
+    // We start reading from the beginning of this chunk, where the offset
+    // equals to chunk_index * common::bytes_per_mb
+    read_file_chunk_request.set_offset_start(chunk_index * 
+                                                 common::bytes_per_mb);
+    read_file_chunk_request.set_length(nbytes);
+    // Access the client-end-point for contacting the chunk server
+    auto server_address(location.server_hostname()+":" + 
+                            std::to_string(location.server_port()));
+    // Prepare the client context 
+    grpc::ClientContext client_context;  
+    common::SetClientContextDeadline(client_context, config_manager_); 
+    // Issue ReadFileChunkRequest and check status
+    StatusOr<ReadFileChunkReply> read_file_chunk_reply_or(
+        chunk_server_service_client_[server_address]->SendRequest(
+            read_file_chunk_request, client_context));
+    
+    // Handle grpc error, log and continue to try the next chunk server
+    if (!read_file_chunk_reply_or.ok()) {
+      LOG(ERROR) << "Read file chunk " << chunk_handle << " from "
+                 << server_address << " failed due to " 
+                 << read_file_chunk_reply_or.status().error_message();
+      continue;
+    }
+
+    // Handle chunk status error, log and continue
+    auto read_file_chunk_reply(read_file_chunk_reply_or.ValueOrDie());
+    switch (read_file_chunk_reply.status()) {
```

Comment:
these error messages are only useful if you also append the chunk server address, so we know which chunk sever failed this request